llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8701,95 +8734,341 @@ SDValue SystemZTargetLowering::combineSETCC(
+                                          int &CCMask,
+                                          DAGCombinerInfo &DCI) const {
   // We have a SELECT_CCMASK or BR_CCMASK comparing the condition code
   // set by the CCReg instruction using the CCValid / CCMask masks,
-  // If the CCReg instruction is itself a ICMP testing the condition
+  // If the CCReg instruction is itself a ICMP / TM  testing the condition
   // code set by some other instruction, see whether we can directly
   // use that condition code.
-
-  // Verify that we have an ICMP against some constant.
-  if (CCValid != SystemZ::CCMASK_ICMP)
-    return false;
-  auto *ICmp = CCReg.getNode();
-  if (ICmp->getOpcode() != SystemZISD::ICMP)
-    return false;
-  auto *CompareLHS = ICmp->getOperand(0).getNode();
-  auto *CompareRHS = dyn_cast<ConstantSDNode>(ICmp->getOperand(1));
-  if (!CompareRHS)
+  auto *CCNode = CCReg.getNode();
+  if (!CCNode)
     return false;
+  const auto getConstFromConstSDVals = [](const SmallVector<SDValue, 4> &Vals) {
+    SmallVector<int, 4> CCVals;
+    for (const auto &Val : Vals)
+      if (auto *ConstNode = dyn_cast<ConstantSDNode>(Val.getNode()))
+        CCVals.emplace_back(ConstNode->getZExtValue());
+      else
+        return SmallVector<int, 4>();
+    return CCVals;
+  };
+  const auto getMSBPosSet = [](unsigned int Mask) {
+    int NumBits = std::numeric_limits<unsigned int>::digits;
+    int count = 0;
+    // Keep target search space to the left.
+    while (NumBits > 0) {
+      NumBits /= 2;
+      // Upper half zeros.
+      if (!(Mask >> NumBits)) {
+        count += NumBits;
+        // Search lower half.
+        Mask <<= NumBits;
+      }
+    }
+    return count;
+  };
 
-  // Optimize the case where CompareLHS is a SELECT_CCMASK.
-  if (CompareLHS->getOpcode() == SystemZISD::SELECT_CCMASK) {
-    // Verify that we have an appropriate mask for a EQ or NE comparison.
-    bool Invert = false;
-    if (CCMask == SystemZ::CCMASK_CMP_NE)
-      Invert = !Invert;
-    else if (CCMask != SystemZ::CCMASK_CMP_EQ)
+  if (CCNode->getOpcode() == SystemZISD::TM) {
+    if (CCValid != SystemZ::CCMASK_TM)
       return false;
-
-    // Verify that the ICMP compares against one of select values.
-    auto *TrueVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(0));
-    if (!TrueVal)
+    const auto emulateTMCCMask = [&](int CCVal, int Mask) {
+      if (!Mask)
+        return std::numeric_limits<unsigned int>::digits;
+      int Result = CCVal & Mask;
+      bool AllOnes = Result == Mask;
+      bool AllZeros = Result == 0;
+      bool MixedZerosOnes = (!AllOnes && !AllZeros);
+      int MSBPos = getMSBPosSet(static_cast<unsigned int>(Mask));
+      bool IsLeftMostBitSet = (Result & (1 << MSBPos)) != 0;
+      return AllOnes                                ? 3
+             : AllZeros                             ? 0
+             : (MixedZerosOnes && IsLeftMostBitSet) ? 2
```

Comment:
According to the sequence of comparisons, `MixedZerosOnes` must be true if we get here, so it seem redundant to test.